### PR TITLE
refactor(app-shell-odd): overhaul odd system update

### PR DESCRIPTION
**Pull request by sfoster1**
Flex robots are capable of updating their own software from the Internet or from USB sticks containing update files. This functionality is handled by the ODD's node side in `app-shell-odd`. It was implemented early in the flex development life cycle before we'd gotten a strong understanding of how everything would work, and was therefore a little scattershot and not really conformant to the way we now think about and structure our code. It also had some unexpected behaviors and was generally a little buggy.

This PR refactors `app-shell-odd`'s system update mechanisms, ideally fixing said bugs.

### changing the frontend or the redux interaction patterns
this was already a pretty big pr, so I didn't touch these. They probably should be touched, though; sourcing updates from web or from usb is a different thing from sourcing updates from web or from the user's file browser, which is what the UI side and the redux messages are designed to implement, and the concepts don't map terribly well. 

The redux message flow and semantics should not change in this PR, and this PR doesn't touch any app code

### changing the precedence or user-visible functionality of system updates
We still have this idea that
- Updates should be presented when an available update is different than the current system version
- Updates from USB sticks have higher precedence than updates from the internet, even if the update from the internet has a higher version than the update from the USB stick
- If multiple updates from sources of the same precedence are available (i.e., a USB stick with multiple update files is connected) then the update with the highest version takes precedence

Changing this would require changing the frontend to have more than a "press a button to start an available update" type of flow, and require changing the redux messages.

### make the code easier to reason about
The code was split sort of weirdly between a single ts file and a module of ts files; code to handle USB updates was scattered around the code to handle internet updates; it was generally a tough assignment to figure out what would happen when user actions are taken or redux messages are sent. Make it easier to understand
### make the code more robust
The code didn't really uniformly handle errors or handle errors in the same way in multiple places; this is pretty much guaranteed because the code didn't really have tests.

## changes
- move all the system update code into `system-update`
- separate the code strongly by layer. the layer that handles the redux messages is in `system-update/handler.ts`, and it imports functions from lower levels in the `from-usb` and `from-web` modules.
- separate the code strongly by source. all code dealing with updates from the web is in the `system-updates/from-web` moduels; the code dealing with updates from usb is in `system-updates/from-usb`. Each module presents a closure with a well-defined interface that the handler can use.
- remove global mutable state in favor of closed-over mutable state for better testability 
- add tests, for everything (this is where the size of the diff is from)

## review requests
- this is a pr that is mostly motivated by readability and robustness, so opinions on how readable and robust the code is are very welcome and needed
- is the behavior correct from the description above?

## reviewing
This ended up being a pretty big PR, sorry! I think a good way to approach looking at is to structure into the mostly-separate changes, which only affect each other implicitly:
- the new cancellation mechanics in HTTP as its own thing
- the action distribution changes in `main`
- the `app.getPath()` changes in `main` and in `config`
- the mechanics of `system-update/handler` first, then the two update providers underneath; the logic here is separated into "conveying the status of updates" in `handler.ts` and "getting updates" in the providers
- the usb device sensing changes, which handle mass storage devices that are like `/dev/sda` but mounted at `/media/VOLUMENAME-sda` and fix excessive timeouts in the recursive enumeration

## testing
this needs to be tested on a flex (conveniently, the OT-2 doesn't run this code). things that should be tested are
- [x] when the robot is connected to the internet on an update channel that has an available update that is higher than the current robot version, does the system download, present, and apply the update
- [x] when the robot is connected to the internet on an update channel that has an available update that is lower than or the same as the current robot version, does the system download but not present the update
- [x] when the robot is connected to the internet on an update channel that has an available update, then disconnected after downloading the available update, does the system present and apply the update
- [x] when the robot is disconnected from the internet while downloading an available update, does the system not present the update
- [x] when you change the update channel, does the presented update get properly cleared and re-queried, including
   - [x] changing from a channel with an update to a channel with a different update
   - [x] changing from a channel with an update to a channel without an update
   - [x] changing from a channel without an update to a channel with an update
- [x] when you plug in a usb stick with an update, it gets detected, presented, and applied
   - [x] including when an update is available from the internet
   - [x] including when the internet update is of a higher version than the usb update
   - [x] including when there are multiple updates on the usb stick
   - [x] including when the update on the usb stick is a lower version than the current version of the robot

Closes RQA-3060

### CODE REVIEW

File: app-shell-odd/src/early.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// things intended to execute early in app-shell initialization
+// do as little as possible in this file and do none of it at import time
+
+import { app } from 'electron'
+import { ODD_DATA_DIR } from './constants'
+
+let path: string
+
+export const setUserDataPath = (): string => {
```
Comment (brenthagen) [inline app-shell-odd/src/early.ts:9]: do we care to split out `get` and `set` functions?

File: app-shell-odd/src/main.ts [hunk 2]
```diff
@@ -87,12 +85,14 @@ function startUp(): void {
   log.info('Starting App')
   console.log('Starting App')
   const storeNeedsReset = fse.existsSync(
-    path.join(ODD_DIR, `_CONFIG_TO_BE_DELETED_ON_REBOOT`)
+    path.join(app.getPath('userData'), `_CONFIG_TO_BE_DELETED_ON_REBOOT`)
```
Comment (brenthagen) [inline app-shell-odd/src/main.ts]: `setUserDataPath()`?

File: app-shell-odd/src/main.ts [hunk 3]
```diff
@@ -102,27 +102,29 @@ function startUp(): void {
 
   // wire modules to UI dispatches
   const dispatch: Dispatch = action => {
-    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
-    if (mainWindow) {
-      log.silly('Sending action via IPC to renderer', { action })
-      mainWindow.webContents.send('dispatch', action)
-    }
+    process.nextTick(() => {
```
Comment (brenthagen) [inline app-shell-odd/src/main.ts:112]: this is probably worth a comment, both the nextTick and the wiring to app-shell handlers

File: app-shell-odd/src/system-update/types.ts [hunk 4]
```diff
@@ -37,3 +37,48 @@ export interface UserFileInfo {
   // parsed contents of VERSION.json
   versionInfo: VersionInfo
```
Comment (brenthagen) [inline app-shell-odd/src/system-update/types.ts]: believe it or not, `UserFileInfo` and `VersionInfo` above is dead code

File: app-shell-odd/src/system-update/handler.ts [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+  reload: () => Promise<void>
+  shouldReload: () => boolean
+  teardown: () => Promise<void>
+}
+
+export function registerUpdateDriver(upstreamDispatch: Dispatch): UpdateDriver {
+  log.info(`Running robot system updates storing to ${getSystemUpdateDir()}`)
+  const dispatch: Dispatch = action => {
+    upstreamDispatch(action)
+  }
```
Comment (brenthagen) [inline app-shell-odd/src/system-update/handler.ts]: why is this needed?

File: app-shell-odd/src/system-update/handler.ts [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+}
+
+export function manageDriver(dispatch: Dispatch): UpdatableDriver {
+  let updateDriver: UpdateDriver | null = null
+  return {
+    handleAction: action => {
+      if (action.type === CONFIG_INITIALIZED) {
+        log.info('Initializing update driver')
+        return new Promise(resolve => {
+          updateDriver = registerUpdateDriver(dispatch)
```
Comment (brenthagen) [inline app-shell-odd/src/system-update/handler.ts]: it's a bit unexpected to nest a `register` within a `register`

File: app-shell-odd/src/actions.ts
Comment (mjhuff) [inline app-shell-odd/src/actions.ts:1]: Two Qs:
- I notice that there are a lot of unused actions here. Is there any reason we're keeping those around?
- For intra-shell communication, are we still keeping the pattern of sending this to browser, and then sending this back to the shell via `meta: {shell: true}`?

File: app-shell-odd/src/system-update/from-usb/scan-zip.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        try {
+          const parsedContents = JSON.parse(contents)
+          if (parsedContents?.robot_type !== 'OT-3 Standard') {
+            reject(new Error('not a Flex release file'))
+          }
+          const fileVersion = parsedContents?.opentrons_api_version
+          const version = Semver.valid(fileVersion as string)
+          if (version === null) {
+            reject(new Error(`${fileVersion} is not a valid version`))
+            return prev
```
Comment (mjhuff) [inline app-shell-odd/src/system-update/from-usb/scan-zip.ts:64]: Should this just be `throw err`?

File: app-shell-odd/src/system-update/from-usb/scan-zip.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        } catch (err: any) {
+          if (err instanceof Error) {
+            log.error(
+              `Failed to read ${entry.name}: ${err.name}: ${err.message}`
+            )
+          } else {
+            log.error(`Failed to ready ${entry.name}: ${err}`)
+          }
+          reject(err)
+          return prev
```
Comment (mjhuff) [inline app-shell-odd/src/system-update/from-usb/scan-zip.ts:79]: See other comment

File: app-shell-odd/src/system-update/handler.ts [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+              resolve()
+            })
+          }
+          usbProviders[action.payload.rootPath] = getUsbUpdateProvider({
+            currentVersion: CURRENT_SYSTEM_VERSION,
+            massStorageDeviceRoot: action.payload.rootPath,
+            massStorageDeviceFiles: action.payload.filePaths,
+          })
+          return usbProviders[action.payload.rootPath]
+            .refreshUpdateCache(() => {})
```
Comment (brenthagen) [inline app-shell-odd/src/system-update/handler.ts:164]: no need for usb progress callback?

File: app-shell-odd/src/system-update/from-web/latest-update.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+export const latestVersionForChannel = (
+  availableVersions: string[],
+  channel: string
+): string | null =>
+  availableVersions
+    .filter(version => channelFinder(version, channel))
+    .sort((a, b) => (semver.gt(a, b) ? 1 : -1))
+    .pop() ?? null
```
Comment (mjhuff) [inline app-shell-odd/src/system-update/from-web/latest-update.ts:20]: I think semver has a built-in method for this:

```
export const latestVersionForChannel = (
  availableVersions: string[],
  channel: string
): string | null => {
  const filteredVersions = availableVersions.filter(version => channelFinder(version, channel));
  return filteredVersions.length > 0 
    ? semver.maxSatisfying(filteredVersions, '*') 
    : null;
};
```

File: app-shell-odd/src/usb.ts [hunk 11]
```diff
@@ -48,11 +54,15 @@ const isWeirdDirectoryAndShouldSkip = (dirName: string): boolean =>
     .map(keyword => dirName.includes(keyword))
     .reduce((prev, current) => prev || current, false)
 
-const enumerateMassStorage = (path: string): Promise<string[]> => {
+const doEnumerateMassStorage = (
+  path: string,
+  depth: number
```
Comment (mjhuff) [inline app-shell-odd/src/usb.ts:59]: Do you think it's worth specifying a max depth?

**Review by mjhuff — APPROVED** [review]
Sweet! I do like the overall structure of this much more. Left some comments for my own edification and potentially (un)helpful things.

**Review by brenthagen — APPROVED** [review]
great leap forward. organization makes sense. main comment, this pushes the boundaries of what we should do without formalized app-shell state management both for scoped values and side effects - at some point we should consider options

File: app-shell-odd/src/actions.ts
Comment (sfoster1) [reply app-shell-odd/src/actions.ts:1]: - afraid to change them, plus some are used by the app-shell because they go through the app common point
- yes but mostly because I really don't want to accidentally break that stuff

File: app-shell-odd/src/system-update/from-usb/scan-zip.ts (hunk 7 above)
Comment (sfoster1) [reply app-shell-odd/src/system-update/from-usb/scan-zip.ts:64]: I dunno. It would mean we don't have to `return`, but I feel gross about `throw`ing in a `new Promise` callback because `reject` is defined and right there...

File: app-shell-odd/src/system-update/from-usb/scan-zip.ts (hunk 8 above)
Comment (sfoster1) [reply app-shell-odd/src/system-update/from-usb/scan-zip.ts:79]: same answer

File: app-shell-odd/src/usb.ts (hunk 11 above)
Comment (sfoster1) [reply app-shell-odd/src/usb.ts:59]: Probably... in a followup though.

File: app-shell-odd/src/system-update/from-web/latest-update.ts (hunk 10 above)
Comment (sfoster1) [reply app-shell-odd/src/system-update/from-web/latest-update.ts:20]: Ooh that's nice. Can I do it in a followup though? I want to get this merged

File: app-shell-odd/src/system-update/handler.ts (hunk 5 above)
Comment (sfoster1) [reply app-shell-odd/src/system-update/handler.ts]: Oh I had some debug logging here that I deleted, I'll flatten this

File: app-shell-odd/src/early.ts (hunk 1 above)
Comment (sfoster1) [reply app-shell-odd/src/early.ts:9]: I guess I could call this `getEnsuredUserDataPath` or something?

File: app-shell-odd/src/system-update/handler.ts (hunk 9 above)
Comment (sfoster1) [reply app-shell-odd/src/system-update/handler.ts:164]: No, because the USB machinery doesn't actually move anything. In theory we could add it for scanning, but we didn't have it before (and quite frankly, we have no UI on the ODD to do anything with the web progress either haha).

File: app-shell-odd/src/system-update/from-web/latest-update.ts (hunk 10 above)
Comment (mjhuff) [reply app-shell-odd/src/system-update/from-web/latest-update.ts:20]: Yeah totally!

File: app-shell-odd/src/system-update/handler.ts (hunk 6 above)
Comment (sfoster1) [reply app-shell-odd/src/system-update/handler.ts]: made it `createUpdateDriver`

**Review by mjhuff — APPROVED** [review]
Post review changes look good, thanks!
